frontends/infinite-corridor/src/api/learning-resources/urls.ts

```diff
@@ -1,54 +1,174 @@
-  options?: Opts
-) => {
-  const { limit = DEFAULT_PAGE_SIZE, offset = 0, ...others } = options ?? {}
-  const searchParams = toQueryString({ limit, offset, ...others })
-  return `/${type}s/?${searchParams}`
+const podcastEpisodeApi = UrlAssembler("/podcastepisodes/")
+const podcastEpisodeDetailsApi = podcastEpisodeApi.segment(":id/")
+const podcastEpisodeUrls = {
+  details: (id: number) => podcastEpisodeDetailsApi.param({ id }).toString(),
+  listing: (options: PaginationSearchParams = {}) =>
+    podcastEpisodeApi
+      .param({ ...DEFAULT_PAGINATION_PARAMS, ...options })
+      .toString()
 }
-const userListItems = (id: number, options?: PaginationSearchParams) => {
-  const { limit = DEFAULT_PAGE_SIZE, offset = 0, ...others } = options ?? {}
-  const searchParams = toQueryString({ limit, offset, ...others })
-  return `/userlists/${id}/items?${searchParams}`
+
+const favoriteApi = UrlAssembler("/favorites/")
+const favoriteUrls = {
+  listing: (options: PaginationSearchParams = {}) =>
+    favoriteApi.param({ ...DEFAULT_PAGINATION_PARAMS, ...options }).toString()
 }
 
+const userListApi = UrlAssembler("/userlists/")
+const userListDetailApi = userListApi.segment(":id/")
+const userListItemsApiU = userListDetailApi.segment("items/")
+const userListItemsDetailApi = userListItemsApiU.segment(":itemId/")
+
 type UserListOptions = { public?: boolean } & PaginationSearchParams
-const userListsListing = (opts?: UserListOptions) =>
-  resourceListing("userlist", opts)
+const userListUrls = {
+  details: (id: number) => userListDetailApi.param({ id }).toString(),
+  create:  userListApi.toString(),
+  listing: (options: UserListOptions = {}) =>
+    userListApi.param({ ...DEFAULT_PAGINATION_PARAMS, ...options }).toString(),
+  itemsListing: (id: number, options: PaginationSearchParams = {}) =>
+    userListItemsApiU
+      .param({ id, ...DEFAULT_PAGINATION_PARAMS, ...options })
+      .toString(),
+  itemDetails: (id: number, itemId: number) =>
+    userListItemsDetailApi.param({ id, itemId }).toString()
+}
 
-const keys = {
-  resourceDetails: (type: string, id: number) => [type, id, "details"],
-  resourceListing: (type: string, options?: PaginationSearchParams) => {
-    if (options === undefined) {
-      return [type, "listing"]
+const resourceUrls = {
+  details: (type: string, id: number) => {
+    switch (type) {
+    case LRT.Course:
+      return courseUrls.details(id)
+    case LRT.Program:
+      return programUrls.details(id)
+    case LRT.Video:
+      return videoUrls.details(id)
+    case LRT.Podcast:
+      return podcastUrls.details(id)
+    case LRT.PodcastEpisode:
+      return podcastEpisodeUrls.details(id)
+    case LRT.Userlist:
+      return userListUrls.details(id)
+    case LRT.LearningPath:
+      return userListUrls.details(id) // LearningPaths are handled by UserList api
+    default:
+      /**
+         * This should not happen and will likely result in a 404.
+         */
+      return `/${type}s/${id}/`
```

Comment:
Should this throw an error instead so we can catch a problem before it when it happens?

Reply:
That was my initial thought, too. On the other hand, the queries are sometimes constructed dynamically from URLs, e.g., the learning resource drawer uses this query: https://discussions-rc.odl.mit.edu/infinite/fields/philosophy/?resource_id=10973&resource_type=course

So if a user copy-pastes the URL wrong  (https://discussions-rc.odl.mit.edu/infinite/fields/philosophy/?resource_id=10973&resource_type=cours), probably the correct behavior is a 404. How should that come about?

**Option 1:** The drawer page validates the URL, and shows a custom error message if it's wrong.

**Option 2:** Just try and make the request, and if it's a 404, show a "resource not found" error message.

What do you think is better? I'm not sure, but I leaned toward Option 2.

**Now:**
- none of that reasoning is clear from the function, and even if we do prefer Option 2, some comment is probably worthwhile
- At the moment, Option 2 doesn't really behave right. The resource drawer has no logic to handle failed queries. It should have that logic.

Reply:
It looks like we only use this in 2 places: tests and the `useResource` hook. It's been a while since I did JS, but if we did throw an error, would this work?

```ts
const useResource = (type: string, id: number) => {
  const key = keys.resource(type).id(id).details
  return useQuery<LearningResource>(key, async () =>{
    const url = urls.resource.details(type, id)
    const result = await axios.get(url)
    return result.data
  })
}
```
Moving `resource.details()` into the callback and making that callback an `async` would mean if the error gets thrown, it'll be a rejected promise and then `isError` on the return value of `useQuery` would reflect that. While the component doesn't handle errors at least the interface would behave correctly.

Reply:
Great idea! Works well. I added that to this PR, but did NOT add a test for it. I added tests for these react-query hooks in the followup branch, so I just added the test to that: https://github.com/mitodl/open-discussions/blob/cc/populate-userlists/frontends/infinite-corridor/src/api/learning-resources/hooks.test.tsx#L53

It turned out this change broke a few tests (exposed an issue, I guess): The `makeLearningResource` factory was creating resources with type `"favorite"`, which was then triggering requests like `/favorites/1234`, which is invalid. "favorites" is fairly separate from the other resource types, so I [separated it out](https://github.com/mitodl/open-discussions/pull/3835/files#diff-dd3d4732baaaff09891b00335b33b3ac776c3ccf3a72d60c06e094e81c9d60d8R23).